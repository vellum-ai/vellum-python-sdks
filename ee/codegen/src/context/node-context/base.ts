import { VellumVariableType } from "vellum-ai/api";

import {
  VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH,
  VELLUM_WORKFLOW_NODES_MODULE_PATH,
} from "src/constants";
import { WorkflowContext } from "src/context";
import { PortContext } from "src/context/port-context";
import { NodeOutputNotFoundError } from "src/generators/errors";
import { NodePort as NodePortType, WorkflowDataNode } from "src/types/vellum";
import { toValidPythonIdentifier } from "src/utils/casing";
import { getNodeLabel } from "src/utils/nodes";
import {
  doesModulePathStartWith,
  getGeneratedNodeDisplayModulePath,
  getGeneratedNodeModuleInfo,
} from "src/utils/paths";
import { hasErrorOutput } from "src/utils/typing";

export declare namespace BaseNodeContext {
  interface Args<T extends WorkflowDataNode> {
    workflowContext: WorkflowContext;
    nodeData: T;
    importOrder: readonly number[];
  }
}

export abstract class BaseNodeContext<T extends WorkflowDataNode> {
  public readonly workflowContext: WorkflowContext;
  public readonly importOrder: readonly number[];
  public readonly nodeModulePath: string[];
  public readonly nodeModuleName: string;
  public readonly nodeFileName: string;
  public readonly nodeClassName: string;
  public readonly nodeDisplayModulePath: string[];
  public readonly nodeDisplayModuleName: string;
  public readonly nodeDisplayFileName: string;
  public readonly nodeDisplayClassName: string;

  public nodeData: T;
  public abstract readonly baseNodeClassName: string;
  public abstract readonly baseNodeDisplayClassName: string;

  protected isCore = false;
  public readonly baseNodeClassModulePath: readonly string[];
  public readonly baseNodeDisplayClassModulePath: readonly string[];

  private nodeOutputNamesById: Record<string, string> | undefined;
  public readonly portContextsById: Map<string, PortContext>;
  public readonly defaultPortContext: PortContext | undefined;

  private readonly nodePortNames: Set<string> = new Set();

  constructor(args: BaseNodeContext.Args<T>) {
    this.workflowContext = args.workflowContext;
    this.nodeData = args.nodeData;
    this.importOrder = args.importOrder;

    if (args.nodeData.type === "GENERIC") {
      if (
        !doesModulePathStartWith(
          args.nodeData.base?.module,
          VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH
        )
      ) {
        this.baseNodeClassModulePath = args.nodeData.base.module;
      } else {
        this.baseNodeClassModulePath = VELLUM_WORKFLOW_NODES_MODULE_PATH;
      }
    } else if (
      !doesModulePathStartWith(
        args.nodeData.base?.module,
        VELLUM_WORKFLOW_NODES_MODULE_PATH
      )
    ) {
      this.baseNodeClassModulePath =
        args.nodeData.base?.module ??
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    } else {
      this.baseNodeClassModulePath =
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    }
    this.baseNodeDisplayClassModulePath =
      this.workflowContext.sdkModulePathNames.NODE_DISPLAY_MODULE_PATH;

    const { moduleName, nodeClassName, modulePath } =
      getGeneratedNodeModuleInfo({
        workflowContext: args.workflowContext,
        nodeDefinition: args.nodeData.definition,
        nodeLabel: this.getNodeLabel(),
      });
    this.nodeModuleName = moduleName;
    this.nodeClassName = nodeClassName;
    this.nodeModulePath = modulePath;

    this.workflowContext.addUsedClassName(nodeClassName);
    this.nodeFileName = `${this.nodeModuleName}.py`;

    this.nodeDisplayModuleName = this.nodeModuleName;
    this.nodeDisplayFileName = `${this.nodeDisplayModuleName}.py`;
    this.nodeDisplayClassName = `${this.nodeClassName}Display`;
    this.nodeDisplayModulePath = getGeneratedNodeDisplayModulePath(
      args.workflowContext,
      this.nodeDisplayModuleName
    );

    const portContexts = this.nodeData.ports
      ? this.createPortContextsFromNodePortAttribute(this.nodeData.ports)
      : this.createPortContexts();
    portContexts.forEach((portContext) => {
      this.workflowContext.addPortContext(portContext);
    });

    this.portContextsById = new Map(
      portContexts.map((portContext) => [portContext.portId, portContext])
    );

    this.defaultPortContext = portContexts.find(
      (portContext) => portContext.isDefault
    );
  }

  protected abstract getNodeOutputNamesById(): Record<string, string>;
  protected abstract getNodeOutputTypesById(): Record<
    string,
    VellumVariableType
  >;
  protected abstract createPortContexts(): PortContext[];

  protected createPortContextsFromNodePortAttribute(
    nodePorts: NodePortType[]
  ): PortContext[] {
    return nodePorts.map(
      (port) =>
        new PortContext({
          workflowContext: this.workflowContext,
          nodeContext: this,
          portId: port.id,
          portName: port.name,
          isDefault: port.type === "DEFAULT",
        })
    );
  }

  public getNodeLabel(): string {
    return getNodeLabel(this.nodeData);
  }

  public getNodeOutputNameById(outputId: string): string | undefined {
    // Lazily load node output names
    if (!this.nodeOutputNamesById) {
      this.nodeOutputNamesById = this.getNodeOutputNamesById();
    }

    const nodeOutputName = this.nodeOutputNamesById[outputId];
    if (!nodeOutputName) {
      // This handles an edge case where the node references a non-existent subworkflow deployment
      if (
        this.nodeData.type === "SUBWORKFLOW" &&
        this.nodeData.data.variant === "DEPLOYMENT"
      ) {
        this.workflowContext.addError(
          new NodeOutputNotFoundError(
            `Could not find Subworkflow Deployment Output with id ${outputId}`,
            "WARNING"
          )
        );
        return;
      }

      this.workflowContext.addError(
        new NodeOutputNotFoundError(
          `Failed to find ${this.nodeClassName} Output with id '${outputId}'`,
          "WARNING"
        )
      );
      return;
    }

    return toValidPythonIdentifier(nodeOutputName, "output");
  }

  public getNodeOutputTypeById(
    outputId: string
  ): VellumVariableType | undefined {
    const nodeOutputTypesById = this.getNodeOutputTypesById();

    return nodeOutputTypesById[outputId];
  }

  public isPortNameUsed(portName: string): boolean {
    return this.nodePortNames.has(portName);
  }

  public addUsedPortName(portName: string): void {
    this.nodePortNames.add(portName);
  }

  /**
   * Extend this class to perform any additional property generation asynchronously
   * after the node context has been created.
   */
  public buildProperties(): Promise<void> {
    return Promise.resolve();
  }

  public isImportedBefore(
    nodeContext: BaseNodeContext<WorkflowDataNode>
  ): boolean {
    for (
      let i = 0;
      i < Math.max(this.importOrder.length, nodeContext.importOrder.length);
      i++
    ) {
      const thisImportOrder = this.importOrder[i];
      const otherImportOrder = nodeContext.importOrder[i];
      if (thisImportOrder === undefined && otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder === undefined) {
        return true;
      }

      if (otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder < otherImportOrder) {
        return true;
      }

      if (thisImportOrder > otherImportOrder) {
        return false;
      }
    }
    return false;
  }

  protected getErrorOutputId(): string | undefined {
    const TRY_NODE_NAME = "TryNode";
    const tryNodes = this.nodeData.adornments?.filter((adornment) => {
      return adornment.base.name === TRY_NODE_NAME;
    });

    const tryNode = tryNodes?.at(-1);

    let fallbackErrorOutputId: string | undefined = undefined;
    if (hasErrorOutput(this.nodeData)) {
      fallbackErrorOutputId = this.nodeData.data.errorOutputId;
    }

    return tryNode ? tryNode.id : fallbackErrorOutputId;
  }
}
