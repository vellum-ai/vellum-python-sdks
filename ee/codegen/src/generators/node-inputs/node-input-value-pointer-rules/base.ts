import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  IterableConfig,
  NodeInputValuePointerRule as NodeInputValuePointerRuleType,
  WorkflowDataNode,
} from "src/types/vellum";

export declare namespace BaseNodeInputValuePointerRule {
  export interface Args<T extends NodeInputValuePointerRuleType> {
    nodeContext: BaseNodeContext<WorkflowDataNode>;
    nodeInputValuePointerRule: T;
    iterableConfig?: IterableConfig;
  }
}

export abstract class BaseNodeInputValuePointerRule<
  T extends NodeInputValuePointerRuleType
> extends AstNode {
  private readonly nodeContext: BaseNodeContext<WorkflowDataNode>;
  public readonly workflowContext: WorkflowContext;
  public readonly nodeInputValuePointerRule: T;
  public readonly iterableConfig?: IterableConfig;
  private astNode: AstNode | undefined;

  constructor({
    nodeContext,
    nodeInputValuePointerRule,
    iterableConfig,
  }: BaseNodeInputValuePointerRule.Args<T>) {
    super();
    this.nodeContext = nodeContext;
    this.workflowContext = nodeContext.workflowContext;
    this.iterableConfig = iterableConfig;
    this.nodeInputValuePointerRule = nodeInputValuePointerRule;

    this.astNode = this.getAstNode();
    if (this.astNode) {
      this.inheritReferences(this.astNode);
    }
  }

  public getReferencedNodeContext():
    | BaseNodeContext<WorkflowDataNode>
    | undefined {
    return undefined;
  }

  abstract getAstNode(): AstNode | undefined;

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
