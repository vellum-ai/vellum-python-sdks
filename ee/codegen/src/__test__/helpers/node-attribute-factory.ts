import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
