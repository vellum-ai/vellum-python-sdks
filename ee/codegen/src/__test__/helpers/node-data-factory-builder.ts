import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
