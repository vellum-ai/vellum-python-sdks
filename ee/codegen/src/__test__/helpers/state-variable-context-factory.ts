import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
