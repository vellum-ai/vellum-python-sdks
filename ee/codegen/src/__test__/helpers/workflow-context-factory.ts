import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
