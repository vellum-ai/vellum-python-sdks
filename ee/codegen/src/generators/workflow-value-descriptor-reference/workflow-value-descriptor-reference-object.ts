import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import { isNil } from "lodash";

import * as codegen from "src/codegen";
import { WorkflowContext } from "src/context";
import {
  IterableConfig,
  WorkflowValueDescriptor as WorkflowValueDescriptorType,
  WorkflowValueDescriptorReference as WorkflowValueDescriptorReferenceType,
} from "src/types/vellum";
import { assertUnreachable } from "src/utils/typing";
import { isReference } from "src/utils/workflow-value-descriptor";

// This class is used to generate the WorkflowValueDescriptorReference as a class and not as the evaluation of the reference
export declare namespace WorkflowValueDescriptorReferenceObject {
  export interface Args {
    workflowContext: WorkflowContext;
    workflowValueDescriptor: WorkflowValueDescriptorType;
    iterableConfig?: IterableConfig;
  }
}

export class WorkflowValueDescriptorReferenceObject extends AstNode {
  private workflowContext: WorkflowContext;
  private iterableConfig?: IterableConfig;
  public astNode: AstNode | undefined;

  constructor(args: WorkflowValueDescriptorReferenceObject.Args) {
    super();

    this.workflowContext = args.workflowContext;
    this.iterableConfig = args.iterableConfig;

    this.astNode = this.getAstNode(args.workflowValueDescriptor);

    this.inheritReferences(this.astNode);
  }

  private getAstNode(
    workflowValueDescriptor: WorkflowValueDescriptorType | undefined
  ): AstNode | undefined {
    if (isNil(workflowValueDescriptor)) {
      return python.TypeInstantiation.none();
    }
    return this.buildExpression(workflowValueDescriptor);
  }

  private buildExpression(
    workflowValueDescriptor: WorkflowValueDescriptorType | undefined
  ): AstNode {
    if (!workflowValueDescriptor) {
      return python.TypeInstantiation.none();
    }

    // Base case
    if (isReference(workflowValueDescriptor)) {
      const workflowExpressionReference =
        this.buildWorkflowExpressionReferenceObject(workflowValueDescriptor);
      this.inheritReferences(workflowExpressionReference);
      return workflowExpressionReference;
    }

    switch (workflowValueDescriptor.type) {
      case "UNARY_EXPRESSION": {
        const lhs = this.buildExpression(workflowValueDescriptor.lhs);
        return python.instantiateClass({
          classReference: python.reference({
            name: "UnaryWorkflowExpression",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "lhs",
              value: lhs,
            }),
            python.methodArgument({
              name: "operator",
              value: python.reference({
                name: "LogicalOperator",
                modulePath:
                  this.workflowContext.sdkModulePathNames
                    .VELLUM_TYPES_MODULE_PATH,
                attribute: ["COALESCE"],
              }),
            }),
          ],
        });
      }
      case "BINARY_EXPRESSION": {
        const lhs = this.buildExpression(workflowValueDescriptor.lhs);
        const rhs = this.buildExpression(workflowValueDescriptor.rhs);
        return python.instantiateClass({
          classReference: python.reference({
            name: "BinaryWorkflowExpression",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "lhs",
              value: lhs,
            }),
            python.methodArgument({
              name: "rhs",
              value: rhs,
            }),
            python.methodArgument({
              name: "operator",
              value: python.reference({
                name: "LogicalOperator",
                modulePath:
                  this.workflowContext.sdkModulePathNames
                    .VELLUM_TYPES_MODULE_PATH,
                attribute: ["COALESCE"],
              }),
            }),
          ],
        });
      }
      case "TERNARY_EXPRESSION": {
        const base = this.buildExpression(workflowValueDescriptor.base);
        const lhs = this.buildExpression(workflowValueDescriptor.lhs);
        const rhs = this.buildExpression(workflowValueDescriptor.rhs);
        return python.instantiateClass({
          classReference: python.reference({
            name: "TernaryWorkflowExpression",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "base",
              value: base,
            }),
            python.methodArgument({
              name: "lhs",
              value: lhs,
            }),
            python.methodArgument({
              name: "rhs",
              value: rhs,
            }),
            python.methodArgument({
              name: "operator",
              value: python.reference({
                name: "LogicalOperator",
                modulePath:
                  this.workflowContext.sdkModulePathNames
                    .VELLUM_TYPES_MODULE_PATH,
                attribute: ["COALESCE"],
              }),
            }),
          ],
        });
      }
      default:
        assertUnreachable(workflowValueDescriptor);
    }
  }

  private buildWorkflowExpressionReferenceObject(
    workflowValueDescriptor: WorkflowValueDescriptorReferenceType
  ): AstNode {
    const referenceType = workflowValueDescriptor.type;

    switch (referenceType) {
      case "NODE_OUTPUT":
        return python.instantiateClass({
          classReference: python.reference({
            name: "NodeOutputWorkflowReference",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "node_id",
              value: python.TypeInstantiation.str(
                workflowValueDescriptor.nodeId
              ),
            }),
            python.methodArgument({
              name: "node_output_id",
              value: python.TypeInstantiation.str(
                workflowValueDescriptor.nodeOutputId
              ),
            }),
          ],
        });
      case "WORKFLOW_INPUT":
        return python.instantiateClass({
          classReference: python.reference({
            name: "WorkflowInputWorkflowReference",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "input_variable_id",
              value: python.TypeInstantiation.str(
                workflowValueDescriptor.inputVariableId
              ),
            }),
          ],
        });
      case "WORKFLOW_STATE":
        return python.instantiateClass({
          classReference: python.reference({
            name: "WorkflowStateVariableWorkflowReference",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "state_variable_id",
              value: python.TypeInstantiation.str(
                workflowValueDescriptor.stateVariableId
              ),
            }),
          ],
        });
      case "CONSTANT_VALUE":
        return python.instantiateClass({
          classReference: python.reference({
            name: "WorkflowStateVariableWorkflowReference",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "value",
              value: codegen.vellumValue({
                vellumValue: workflowValueDescriptor.value,
                iterableConfig: this.iterableConfig,
              }),
            }),
          ],
        });
      case "VELLUM_SECRET":
        return python.instantiateClass({
          classReference: python.reference({
            name: "VellumSecretWorkflowReference",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "vellum_secret_name",
              value: python.TypeInstantiation.str(
                workflowValueDescriptor.vellumSecretName
              ),
            }),
          ],
        });
      case "EXECUTION_COUNTER":
        return python.instantiateClass({
          classReference: python.reference({
            name: "ExecutionCounterWorkflowReference",
            modulePath:
              this.workflowContext.sdkModulePathNames.VELLUM_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "node_id",
              value: python.TypeInstantiation.str(
                workflowValueDescriptor.nodeId
              ),
            }),
          ],
        });
      default: {
        assertUnreachable(referenceType);
      }
    }
  }

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
