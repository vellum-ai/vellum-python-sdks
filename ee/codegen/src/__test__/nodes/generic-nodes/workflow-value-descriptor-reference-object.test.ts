import { Writer } from "@fern-api/python-ast/core/Writer";
import { DocumentIndexRead } from "vellum-ai/api";
import { DocumentIndexes as DocumentIndexesClient } from "vellum-ai/api/resources/documentIndexes/client/Client";
import { vi } from "vitest";

import {
  nodeContextFactory,
  workflowContextFactory,
} from "src/__test__/helpers";
import { mockDocumentIndexFactory } from "src/__test__/helpers/document-index-factory";
import { searchNodeDataFactory } from "src/__test__/helpers/node-data-factories";
import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import { WorkflowValueDescriptorReferenceObject } from "src/generators/workflow-value-descriptor-reference/workflow-value-descriptor-reference-object";
import {
  WorkflowDataNode,
  WorkflowValueDescriptor as WorkflowValueDescriptorType,
} from "src/types/vellum";

describe("WorkflowValueDescriptorReferenceObject", () => {
  let writer: Writer;
  let workflowContext: WorkflowContext;

  beforeEach(() => {
    writer = new Writer();
    workflowContext = workflowContextFactory();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("should generate correct AST for CONSTANT_VALUE reference", async () => {
    const constantValueDescriptor: WorkflowValueDescriptorType = {
      type: "CONSTANT_VALUE",
      value: {
        type: "STRING",
        value: "Hello, World!",
      },
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: constantValueDescriptor,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for NODE_OUTPUT reference", async () => {
    vi.spyOn(workflowContext, "findNodeContext").mockReturnValue({
      nodeClassName: "TestNode",
      path: ["nodes", "test-node-path"],
      getNodeOutputNameById: vi.fn().mockReturnValue("my_output"),
    } as unknown as BaseNodeContext<WorkflowDataNode>);

    const nodeOutputDescriptor: WorkflowValueDescriptorType = {
      type: "NODE_OUTPUT",
      nodeId: "test-node-id",
      nodeOutputId: "test-output-id",
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: nodeOutputDescriptor,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for VELLUM_SECRET reference", async () => {
    const secretDescriptor: WorkflowValueDescriptorType = {
      type: "VELLUM_SECRET",
      vellumSecretName: "API_KEY",
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: secretDescriptor,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for EXECUTION_COUNTER reference", async () => {
    vi.spyOn(DocumentIndexesClient.prototype, "retrieve").mockResolvedValue(
      mockDocumentIndexFactory() as unknown as DocumentIndexRead
    );
    const node = searchNodeDataFactory();
    await nodeContextFactory({ workflowContext, nodeData: node });
    const counterDescriptor: WorkflowValueDescriptorType = {
      type: "EXECUTION_COUNTER",
      nodeId: node.id,
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: counterDescriptor,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for WORKFLOW_INPUT reference", async () => {
    const inputDescriptor: WorkflowValueDescriptorType = {
      type: "WORKFLOW_INPUT",
      inputVariableId: "test-input-id",
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: inputDescriptor,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for WORKFLOW_STATE reference", async () => {
    const stateDescriptor: WorkflowValueDescriptorType = {
      type: "WORKFLOW_STATE",
      stateVariableId: "test-state-id",
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: stateDescriptor,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for UNARY_EXPRESSION", async () => {
    const unaryExpression: WorkflowValueDescriptorType = {
      type: "UNARY_EXPRESSION",
      operator: "null",
      lhs: {
        type: "CONSTANT_VALUE",
        value: {
          type: "STRING",
          value: "Hello, World!",
        },
      },
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: unaryExpression,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for BINARY_EXPRESSION", async () => {
    const binaryExpression: WorkflowValueDescriptorType = {
      type: "BINARY_EXPRESSION",
      operator: "=",
      lhs: {
        type: "CONSTANT_VALUE",
        value: {
          type: "STRING",
          value: "Hello",
        },
      },
      rhs: {
        type: "CONSTANT_VALUE",
        value: {
          type: "STRING",
          value: "World",
        },
      },
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: binaryExpression,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for TERNARY_EXPRESSION", async () => {
    const ternaryExpression: WorkflowValueDescriptorType = {
      type: "TERNARY_EXPRESSION",
      operator: "between",
      base: {
        type: "CONSTANT_VALUE",
        value: {
          type: "NUMBER",
          value: 5,
        },
      },
      lhs: {
        type: "CONSTANT_VALUE",
        value: {
          type: "NUMBER",
          value: 1,
        },
      },
      rhs: {
        type: "CONSTANT_VALUE",
        value: {
          type: "NUMBER",
          value: 10,
        },
      },
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: ternaryExpression,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for nested BINARY_EXPRESSION with 4 node outputs", async () => {
    vi.spyOn(workflowContext, "findNodeContext").mockReturnValue({
      nodeClassName: "TestNode",
      path: ["nodes", "test-node-path"],
      getNodeOutputNameById: vi.fn().mockReturnValue("my_output"),
    } as unknown as BaseNodeContext<WorkflowDataNode>);

    // Create a nested binary expression with 4 node outputs
    const nestedBinaryExpression: WorkflowValueDescriptorType = {
      type: "BINARY_EXPRESSION",
      operator: "coalesce",
      lhs: {
        type: "BINARY_EXPRESSION",
        operator: "coalesce",
        lhs: {
          type: "BINARY_EXPRESSION",
          operator: "coalesce",
          lhs: {
            type: "NODE_OUTPUT",
            nodeId: "node-id-1",
            nodeOutputId: "output-id-1",
          },
          rhs: {
            type: "NODE_OUTPUT",
            nodeId: "node-id-2",
            nodeOutputId: "output-id-2",
          },
        },
        rhs: {
          type: "NODE_OUTPUT",
          nodeId: "node-id-3",
          nodeOutputId: "output-id-3",
        },
      },
      rhs: {
        type: "NODE_OUTPUT",
        nodeId: "node-id-4",
        nodeOutputId: "output-id-4",
      },
    };

    const referenceObject = new WorkflowValueDescriptorReferenceObject({
      workflowContext,
      workflowValueDescriptor: nestedBinaryExpression,
    });

    referenceObject.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
