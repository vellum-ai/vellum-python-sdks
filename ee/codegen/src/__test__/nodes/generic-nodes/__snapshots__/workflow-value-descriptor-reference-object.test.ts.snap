// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for BINARY_EXPRESSION 1`] = `
"BinaryWorkflowExpression(
    lhs=WorkflowStateVariableWorkflowReference(value="Hello"),
    rhs=WorkflowStateVariableWorkflowReference(value="World"),
    operator=LogicalOperator.COALESCE,
)
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for CONSTANT_VALUE reference 1`] = `
"WorkflowStateVariableWorkflowReference(value="Hello, World!")
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for EXECUTION_COUNTER reference 1`] = `
"ExecutionCounterWorkflowReference(node_id="search")
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for NODE_OUTPUT reference 1`] = `
"NodeOutputWorkflowReference(node_id="test-node-id", node_output_id="test-output-id")
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for TERNARY_EXPRESSION 1`] = `
"TernaryWorkflowExpression(
    base=WorkflowStateVariableWorkflowReference(value=5),
    lhs=WorkflowStateVariableWorkflowReference(value=1),
    rhs=WorkflowStateVariableWorkflowReference(value=10),
    operator=LogicalOperator.COALESCE,
)
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for UNARY_EXPRESSION 1`] = `
"UnaryWorkflowExpression(
    lhs=WorkflowStateVariableWorkflowReference(value="Hello, World!"),
    operator=LogicalOperator.COALESCE,
)
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for VELLUM_SECRET reference 1`] = `
"VellumSecretWorkflowReference(vellum_secret_name="API_KEY")
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for WORKFLOW_INPUT reference 1`] = `
"WorkflowInputWorkflowReference(input_variable_id="test-input-id")
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for WORKFLOW_STATE reference 1`] = `
"WorkflowStateVariableWorkflowReference(state_variable_id="test-state-id")
"
`;

exports[`WorkflowValueDescriptorReferenceObject > should generate correct AST for nested BINARY_EXPRESSION with 4 node outputs 1`] = `
"BinaryWorkflowExpression(
    lhs=BinaryWorkflowExpression(
        lhs=BinaryWorkflowExpression(
            lhs=NodeOutputWorkflowReference(
                node_id="node-id-1", node_output_id="output-id-1"
            ),
            rhs=NodeOutputWorkflowReference(
                node_id="node-id-2", node_output_id="output-id-2"
            ),
            operator=LogicalOperator.COALESCE,
        ),
        rhs=NodeOutputWorkflowReference(
            node_id="node-id-3", node_output_id="output-id-3"
        ),
        operator=LogicalOperator.COALESCE,
    ),
    rhs=NodeOutputWorkflowReference(node_id="node-id-4", node_output_id="output-id-4"),
    operator=LogicalOperator.COALESCE,
)
"
`;
